chia/util/action_scope.py

```diff
@@ -0,0 +1,174 @@
+        await self.get_active_writer().execute("DELETE FROM side_effects")
+        await self.get_active_writer().execute(
+            "INSERT INTO side_effects VALUES(?)",
+            (bytes(resource),),
+        )
+
+    async def get_memos(self) -> Dict[bytes, bytes]:
+        rows = await self.get_active_writer().execute_fetchall("SELECT key, value FROM memos")
+        memos = {row[0]: row[1] for row in rows}
+        return memos
+
+    async def save_memos(self, memos: Dict[bytes, bytes]) -> None:
+        await self.get_active_writer().execute("DELETE FROM memos")
+        await self.get_active_writer().executemany(
+            "INSERT INTO memos VALUES(?, ?)",
+            memos.items(),
+        )
+
+
+class SideEffects(Protocol):
+    def __bytes__(self) -> bytes: ...
+
+    @classmethod
+    def from_bytes(cls: Type[_T_SideEffects], blob: bytes) -> _T_SideEffects: ...
+
+
+_T_SideEffects = TypeVar("_T_SideEffects", bound=SideEffects)
+
+
+@dataclass
+class ActionScope(Generic[_T_SideEffects]):
+    """
+    The idea of a wallet action is to map a single user input to many potentially distributed wallet functions and side
+    effects. The eventual goal is to have this be the only connection between a wallet type and the WSM.
+
+    Utilizes a "resource manager" to hold the state in order to take advantage of rollbacks and prevent concurrent tasks
+    from interferring with each other.
+    """
+
+    _resource_manager: ResourceManager
+    _side_effects_format: Type[_T_SideEffects]
+    _callbacks: List[Callable[[StateInterface[_T_SideEffects]], Awaitable[None]]] = field(default_factory=list)
+    _final_side_effects: Optional[_T_SideEffects] = field(init=False, default=None)
+
+    @property
+    def side_effects(self) -> _T_SideEffects:
+        if self._final_side_effects is None:
+            raise RuntimeError(
+                "Can only request ActionScope.side_effects after exiting context manager. "
+                "While in context manager, use ActionScope.use()."
+            )
+
+        return self._final_side_effects
+
+    @classmethod
+    @contextlib.asynccontextmanager
+    async def new_scope(
+        cls,
+        side_effects_format: Type[_T_SideEffects],
+        resource_manager_backend: Type[ResourceManager] = SQLiteResourceManager,
+    ) -> AsyncIterator[ActionScope[_T_SideEffects]]:
+        async with resource_manager_backend.managed(side_effects_format()) as resource_manager:
+            self = cls(_resource_manager=resource_manager, _side_effects_format=side_effects_format)
+            try:
+                yield self
+            except Exception:
+                raise
+            else:
+                async with self.use(_callbacks_allowed=False) as interface:
+                    for callback in self._callbacks:
+                        await callback(interface)
+                    self._final_side_effects = interface.side_effects
```

Comment:
Is there really nothing to clean up in the case of an exception?

Just from a language mechanics perspective, this can be simplified.

```suggestion
            yield self

            async with self.use(_callbacks_allowed=False) as interface:
                for callback in self._callbacks:
                    await callback(interface)
                self._final_side_effects = interface.side_effects
```

Reply:
No, this is all temporary state while it's active. Only once everything finishes do you decide to commit it if you want.  Your suggestion has been taken: ef59007507

Reply:
Oh actually pylint didn't like this one either

Reply:
Changed back pending pylint fix in main: 94b2d65f53